packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2680,46 +2676,80 @@ export class ContainerRuntime
 				local,
 			);
 			if (messagesWithPendingState.length > 0) {
-				messagesWithPendingState.forEach(({ message, localOpMetadata }) => {
-					const msg: MessageWithContext = {
-						message,
-						local,
-						isRuntimeMessage: true,
-						savedOp,
-						localOpMetadata,
-					};
-					this.ensureNoDataModelChanges(() => this.processRuntimeMessage(msg));
-				});
+				this.batchBegin(messagesWithPendingState[0]?.message);
+
+				let error: any;
+				try {
+					messagesWithPendingState.forEach(({ message, localOpMetadata }) => {
+						const msg: MessageWithContext = {
+							message,
+							local,
+							isRuntimeMessage: true,
+							savedOp,
+							localOpMetadata,
+						};
+						this.ensureNoDataModelChanges(() => this.processRuntimeMessage(msg));
+					});
+				} catch (e) {
+					error = e;
+					throw e;
+				} finally {
+					this.batchEnd(
+						error,
+						messagesWithPendingState[messagesWithPendingState.length - 1]?.message,
+					);
+				}
 			} else {
 				this.ensureNoDataModelChanges(() => this.processEmptyBatch(inboundBatch, local));
```

Comment:
What are your thoughts on why `processEmptyBatch` doesn't call delta scheduler?  To me, it is similar to the `else` case below (legacy or non-runtime message) - we got a single message over the wire (the empty grouped batch, in this case).  So if we notify in that case we should do the same here.

It looks like DeltaScheduler only needs to know the sequenceNumber, so you can just change `message: ISequencedDocumentMessage` in `batchBegin`/`batchEnd` sigature to `{ sequenceNumber }: { sequenceNumber: number }`, similar to what `processEmptyBatch` uses for the event args.

Reply:
I guess it's been that way since @dannimad introduced `processEmptyBatch` - Daniel any input here?  IMO these different codepaths should be as similar as possible (and share as much of that code as possible too)

Reply:
See comment below about suggesting to remove the explicit DeltaScheduler call and have that listen to the events.

Reply:
Also, FYI the empty batch case is not respecting the event args type specified on `IContainerRuntimeBaseEvents`.  The alternative would be to pass the original empty grouped batch (simply `messageCopy`, I believe).  cc @dannimad for this as well.

Reply:
I am not too sure about the empty batch case. It's incorrect that it violates the `IContainerRuntimeBaseEvents` arguments. Anyone listening to this can fail unexpectedly. I would prefer not to emit these events if there was nothing processed.
I would not want to change it in this PR because that is significant change in behavior and I don't know what it's supposed to do.
@dannimad can you please look into this.

Reply:
I believe we should call delta scheduler on the empty batch case as well. We could just make slight changes to current container::batchBegin and batchEnd methods and call them from processEmptyBatch(). 

These side effects were implanted in the empty batch case just to make sure the workflow were as close as the non-empty case as possible, not because we must but out of ignorance to know what could be some unwanted consequences of making that case more unique.

Reply:
If we have a strong reason not to call it, I would also remove emmiting the event from the empty case (which does not sound right to me).

Reply:
I do not have enough context on why we need empty batches (it would be great to have some note somewhere describing why this is needed and how it works), but assuming it's needed, will having runtime noop op (as part of such batch) be a good idea? The metric of "good" for me here - we have less concepts. Partially having noop adds a concept, but not having such things as processEmptyBatch() and special casing here and there might overall reduce number of concepts.

Today, loader will not allow it (if that's good direction - it's easy to change) - would using system noop (i.e. runtime sending "regular" noop) suffice where we leverage today empty batch?

Reply:
We considered adding a runtime no-op, but didn't wnat to wait for saturation blah blah blah.  So we went with an empty grouped batch.  I think we should just project that empty grouped batch op to the rest of the system, no one should be inspecting it anyway.  It would be surprising to anyone who _was_ closely inspecting the op payload, because other than this you'd never expect to see a Grouped Batch at this layer (it would be ungrouped and you'd get actual batch messages themselves in these cases).  But I think it's fine.

Reply:
Several people have asked for info about why empty batches.  There are some comments scattered around wherever we are dealing with empty batches, but that's not enough.

Where did you look for info but not find it?  I'll add it there.

Reply:
Opened AB#13989 about the gaps around empty batches